File: frontend/src/services/commentService.ts
import api from '../api/api';
import type { QuestionComment, ApiResponse } from '../types/types';

// Get comments for a question
export const getQuestionComments = async (questionId: string): Promise<QuestionComment[]> => {
    const response = await api.get(`/questions/${questionId}/comments`);
    return response.data.data;
};

// Add comment to a question
export const addComment = async (questionId: string, commentText: string): Promise<QuestionComment> => {
    const response = await api.post(`/questions/${questionId}/comments`, { comment_text: commentText });
    return response.data.data;
};

// Update a comment
export const updateComment = async (commentId: string, commentText: string): Promise<QuestionComment> => {
    const response = await api.put(`/comments/${commentId}`, { comment_text: commentText });
    return response.data.data;
};

// Delete a comment
export const deleteComment = async (commentId: string): Promise<ApiResponse<null>> => {
    const response = await api.delete(`/comments/${commentId}`);
    return response.data;
};

File: internal/services/comment.service.go
package services

import (
	"context"
	dto_comment "ecoquiz/internal/dto/comment"
	"ecoquiz/internal/models"
	"ecoquiz/internal/repos"
	"errors"

	"github.com/jackc/pgx/v5"
)

type CommentService struct {
	commentRepo  repos.CommentRepo
	questionRepo repos.QuestionRepo
}

func NewCommentService(commentRepo repos.CommentRepo, questionRepo repos.QuestionRepo) *CommentService {
	return &CommentService{
		commentRepo:  commentRepo,
		questionRepo: questionRepo,
	}
}

func (s *CommentService) CreateComment(ctx context.Context, questionID, userID string, req *dto_comment.CreateCommentReq) (string, error) {
	// Check if question exists
	_, err := s.questionRepo.GetByID(ctx, questionID)
	if err != nil {
		if err == pgx.ErrNoRows {
			return "", errors.New("question not found")
		}
		return "", errors.New("failed to get question")
	}

	comment := &models.QuestionComment{
		QuestionID:  questionID,
		UserID:      userID,
		CommentText: req.Text,
	}

	if err := s.commentRepo.Create(ctx, comment); err != nil {
		return "", errors.New("failed to create comment: " + err.Error())
	}
	return comment.ID, nil
}

func (s *CommentService) DeleteComment(ctx context.Context, commentID, userID string) error {
	comment, err := s.commentRepo.GetByID(ctx, commentID)
	if err != nil {
		return err
	}

	if comment.UserID != userID {
		return errors.New("unauthorized: you can only delete your own comments")
	}

	if err := s.commentRepo.Delete(ctx, commentID); err != nil {
		return errors.New("failed to delete comment: " + err.Error())
	}
	return nil
}
